src/derivers/cip3.ts

```diff
@@ -0,0 +1,308 @@
+ * @param param1.childIndex - The index of the child key.
+ * @param param1.isHardened - Indicates if the child key is hardened.
+ * @returns The derived child key.
+ */
+export const derivePrivateKey = async ({
+  parentNode,
+  childIndex,
+  isHardened,
+}: DeriveWithPrivateArgs) => {
+  // extension = i >= 2^31 ? (0x00||kp||i) : (0x02||Ap||i)
+  const extension = isHardened
+    ? getKeyExtension(
+        Z_TAGS.hardened,
+        parentNode.privateKeyBytes,
+        childIndex + BIP_32_HARDENED_OFFSET,
+      )
+    : getKeyExtension(Z_TAGS.normal, parentNode.publicKeyBytes, childIndex);
+
+  // entropy = Fcp(extension)
+  const entropy = generateEntropy({
+    chainCode: parentNode.chainCodeBytes,
+    extension,
+  });
+
+  const zl = entropy.subarray(0, 32);
+  const zr = entropy.subarray(32);
+
+  const parentKl = parentNode.privateKeyBytes.subarray(0, 32);
+  const parentKr = parentNode.privateKeyBytes.subarray(32);
+
+  // 8[ZL] + kPL
+  const childKl = add(trunc28Mul8(zl), parentKl);
+  // ZR + kPR
+  const childKr = add(zr, mod2Pow256(parentKr));
+  return concatBytes([childKl, childKr]);
+};
+
+type DeriveWithoutPrivateArgs = DeriveKeyBaseArgs & {
+  parentNode: {
+    chainCodeBytes: Uint8Array;
+    publicKeyBytes: Uint8Array;
+  };
+  isHardened: false;
+};
+
+const CHAIN_CODE_TAGS = {
+  normal: 3,
+  hardened: 1,
+};
+
+/**
+ * Derives a child chainCode.
+ *
+ * Following "Section V. BIP32-ED25519: SPECIFICATION, C.3" in https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf.
+ *
+ * @param param1 - The parameters for deriving a child chainCode.
+ * @param param1.parentNode - The parent node containing optionally a private key, chain code, and public key.
+ * @param param1.childIndex - The index of the child key.
+ * @param param1.isHardened - Indicates if the child key is hardened.
+ * @returns The derived child chainCode.
+ */
+export const deriveChainCode = async ({
+  parentNode,
+  childIndex,
+  isHardened,
+}: DeriveWithPrivateArgs | DeriveWithoutPrivateArgs) => {
+  // extension = i >= 2^31 ? (0x01||kp||i) : (0x03||Ap||i)
+  const extension = isHardened
+    ? getKeyExtension(
+        CHAIN_CODE_TAGS.hardened,
+        parentNode.privateKeyBytes,
+        childIndex + BIP_32_HARDENED_OFFSET,
+      )
+    : getKeyExtension(
+        CHAIN_CODE_TAGS.normal,
+        parentNode.publicKeyBytes,
+        childIndex,
+      );
+
+  // entropy = Fcp(extension)
+  const entropy = generateEntropy({
+    chainCode: parentNode.chainCodeBytes,
+    extension,
+  });
+
+  return entropy.subarray(32);
+};
+
+export const derivePublicKey = async ({
+  parentNode,
+}: DeriveWithoutPrivateArgs) => {
+  return parentNode.publicKeyBytes; // FIXME: replace with actual derivation
+};
+
+/**
+ * Derive a SLIP-10 child key with a given path from a parent key.
+ *
+ * @param options - The options for deriving a child key.
+ * @returns SLIP10Node.
+ */
+export async function deriveChildKey(
```

Comment:
This `fn` is basically the entrypoint, if the consumer wants to derive cardano keys, he will pass path as `["cip3:1852'", "cip3:1815'"]` and this function then gets called on each path element.